custom_components/tesla_custom/base.py

```diff
@@ -0,0 +1,183 @@
+"""Support for Tesla cars."""
+from teslajsonpy.car import TeslaCar
+from teslajsonpy.energy import EnergySite
+
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+from homeassistant.util import slugify
+
+from . import TeslaDataUpdateCoordinator
+from .const import ATTRIBUTION, DOMAIN
+
+DEFAULT_DEVICE = "device"
+
+
+class TeslaBaseEntity(CoordinatorEntity):
+    """Representation of a Tesla device."""
+
+    _attr_attribution = ATTRIBUTION
+    _attr_has_entity_name = True
+
+    def __init__(
+        self, hass: HomeAssistant, coordinator: TeslaDataUpdateCoordinator
+    ) -> None:
+        """Initialise the Tesla device."""
+        super().__init__(coordinator)
+        self._coordinator = coordinator
+        self.hass = hass
+        # reset the device type. If its not already set, it sets it to the
+        # default device.
+        self.type: str = getattr(self, "type", DEFAULT_DEVICE)
```

Comment:
Why do we need to reset the device type and set it to a default device? From what I can tell, every entity when it's created will have a type set which is required to make sure the name and UUID is correct.

Reply:
The comment is probably wrong that it's "resetting". It's checking if there's a type and will set to default_device if it doesn't have it. Not sure it's required but it's an error check.

Reply:
Given what we're using `self.type` for (name and unique ID) if someone adds new classes later and doesn't specify `self.type`, I think HA will still create the entities but add a `_2` to the entity name and I'm not actually sure what would happen with the unique ID. My concern is this just allows an error (not specifying `self.type`) go unnoticed. Removing this would throw an error when HA tries to run since we use `self.type`, which I think is what we would want.

Reply:
Will that default error break the integration functionality because it's uncaught? You have to choose which battles you want to fight. I'm fine for catching the error and printing a warning to come report, but again, it's luck if someone will report it.

Reply:
I think it would cause some issues since two entities would be sharing the same unique ID. My thought is I rather have it throw an error when someone is developing a new class so it's caught immediately and fixed. This also looks like something @Megabytemb intentionally added with the rewrite.